Feature/debugger detection

ha-ja opened the pull request:
Added new detector:
- Debugger detection added

ha-ja committed:
add DebuggerDetection

ha-ja committed:
add DebuggerDetection to example app

theDeniZ commented on securitytoolkit/src/main/java/com/exxeta/securitytoolkit/internal/DebuggerDetection.kt:
```diff
@@ -0,0 +1,65 @@
+        } catch (e: Exception) {
+            isPortInUse = false
+        }
+        return isPortInUse
```
instead of creating a variable you could do something like that:

```
private fun isPortInUse(port: Int): Boolean = try {
  Socket().use { s -> 
    s.connect(InetSocketAddress("127.0.0.1", port), 200)
  }
  true
} catch (e: Throwable) {
  false
}
```

note the `Throwable` instead of `Exception`

ha-ja replied:
👍 Done

theDeniZ commented on securitytoolkit/src/main/java/com/exxeta/securitytoolkit/internal/DebuggerDetection.kt:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    private fun isAdbPortListening(): Boolean {
+        return adbPorts.any { isPortInUse(it.toInt()) }
```
given `adbPorts` were a set / list of integers already (instead of doing it.toString() at the top), you could write it as `adbPorts.any(isPortInUse)`

ha-ja replied:
👍 Done

theDeniZ commented on securitytoolkit/src/main/java/com/exxeta/securitytoolkit/internal/DebuggerDetection.kt:
```diff
@@ -0,0 +1,65 @@
+            isTracerPidSet = null
+        }
+        return isTracerPidSet
+    }
```
you could safe few lines by
```
 private fun isTracerPidSet(): Boolean? = try {
  FileReader("/proc/self/status").buffered().let { br ->
      br.lines().anyMatch {
          it.startsWith("TracerPid:") && it.split(":")[1].trim() != "0"
      }
  }
} catch (e: Throwable) {
  null
}
```

Note `Throwable` instead of `IOException`

theDeniZ replied:
Thinking about it, it would also be better to return false instead of null, as the /proc/self/status file is not solely responsible for debugger status, thus making here `null` response ambiguous. As Exception here concerns file operations most of the time, it would be inconclusive to say anything about debugger

ha-ja replied:
I agree with you. 
Especially with the latest Android smartphones, this check no longer says much. 
It is also not guaranteed that /proc/self/status can be viewed by every app. This should not lead to the checks being aborted.
Is a nice try and should not be more.
Should now be solved after the last commit.

theDeniZ commented on securitytoolkit/src/main/java/com/exxeta/securitytoolkit/internal/DebuggerDetection.kt:
```diff
@@ -0,0 +1,65 @@
+     */
+    fun threatDetected(): Boolean? {
+        val isTracerPidSet = isTracerPidSet() ?: return null
+        return Debug.isDebuggerConnected() || Debug.waitingForDebugger() || isAdbPortListening() || isTracerPidSet
```
i find this logic a bit counterintuitive: if isTracerPidSet is not null, it will influence the resulted value in the last place, but if it is null, we skip all other checks (which may give conclusive results). Something seems odd

Maybe try it like that?
``` 
if (Debug.isDebuggerConnected() || Debug.waitingForDebugger() || isAdbPortListening()) {
  return true
} else {
  return isTracerPidSet()
}
```

ha-ja replied:
I have simplified it even further. Return type is now a Boolean and no longer a Boolean? .

theDeniZ commented on securitytoolkit/src/main/java/com/exxeta/securitytoolkit/internal/DebuggerDetection.kt:
```diff
@@ -0,0 +1,65 @@
+internal object DebuggerDetection {
+
+    private val adbPorts = (5555..5585 step 2).map { it.toString() }
+    private val adbConsolePorts = (5554..5584 step 2).map { it.toString() }
```
variable not used

ha-ja committed:
simplify checks, remove unused code

- threatDetected() condition fixed and simplified.
- private isTracerPidSet() simplified as a single-expression function.
- private isPortInUse() simplified as a single-expression function.
- removed unnecessary string casting from val adbPorts list.
- removed val adbConsolePorts (not yet used).

theDeniZ reviewed (approved):
looks good
